### PR TITLE
[CDAP-18648] Get stopping timeout from Runtime Handler

### PR DESCRIPTION
**Pull request by snehalakshmisha**
This change enables RuntimeHandler to return the program run status and the stopping timeout if the run status is STOPPING to the RuntimeClient.
Jira - https://cdap.atlassian.net/browse/CDAP-18649

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJob.java [hunk 1]
```diff
@@ -256,6 +257,7 @@ public void error(Throwable cause) {
   @Override
   public void requestStop() {
     try {
+      runtimeClientService.getStoppingCommandWithGracefulPeriod();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJob.java]: Why this is called inside requestStop()?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java [hunk 2]
```diff
@@ -73,48 +73,49 @@
...
-    if (!exists) {
-      throw new BadRequestException("Program run " + programRunId + " is not valid");
-    }
   }
 
-  /**
-   * Checks if the given {@link ProgramRunId} is valid.
-   */
-  private boolean isValid(ProgramRunId programRunId) throws IOException, UnauthorizedException {
+  private RunRecordDetail runRecord(ProgramRunId programRunId) throws IOException, UnauthorizedException,
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java]: Call the method `getRunRecord`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java [hunk 3]
```diff
@@ -73,48 +73,49 @@
...
     // If it is not found in the local store, which should be very rare, try to fetch the run record remotely.
     try {
       LOG.info("Remotely fetching program run details for {}", programRunId);
       runRecord = runRecordFetcher.getRunRecordMeta(programRunId);
       // Try to update the local store
       insertRunRecord(programRunId, runRecord);
-      return !runRecord.getStatus().isEndState();
+      return runRecord;
+    } catch (IOException e) {
+      LOG.error("Exception occurred while trying to fetch run record detail");
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java]: Don't log and throw.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java [hunk 4]
```diff
@@ -73,48 +73,49 @@
...
+  public RunRecordDetail checkRunRecord(ProgramRunId programRunId, HttpRequest request) throws BadRequestException {
     accessEnforcer.enforce(programRunId, authenticationContext.getPrincipal(), StandardPermission.GET);
-    boolean exists;
+    RunRecordDetail runRecord;
     try {
-      exists = programRunsCache.get(programRunId);
+      runRecord = programRunsCache.get(programRunId);
+      return runRecord;
     } catch (Exception e) {
       throw new ServiceUnavailableException(Constants.Service.RUNTIME, e);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java:102]: Why it is service unavailable?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeHandler.java [hunk 5]
```diff
@@ -317,12 +325,12 @@ public void chunk(ByteBuf request, HttpResponder responder) {
     public void finished(HttpResponder responder) {
       try {
         if (payloads.isEmpty()) {
-          responder.sendStatus(HttpResponseStatus.OK);
+          responder.sendJson(HttpResponseStatus.OK, GSON.toJson(runRecord, RunRecordDetail.class));
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeHandler.java]: Why are we sending back the whole run record? We should only send something back if the state is stopping? Also, it is better be a different data structure that carry less data than the whole run record detail.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/ProgramRunStatus.java [hunk 6]
```diff
@@ -33,6 +33,7 @@
   RUNNING,
   SUSPENDED,
   RESUMING,
+  STOPPING,
```
Comment (CuriousVini) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/ProgramRunStatus.java]: how does the program transitions look like if we are adding new state in `canTransitionTo ` method?

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocJobMain.java [hunk 7]
```diff
@@ -140,7 +140,7 @@ public static void main(String[] args) throws Exception {
       } finally {
         // call destroy() method on envProviderClass
         Method closeMethod = dataprocEnvClass.getMethod("destroy");
-        LOG.info("Invoking destroy() on {}", runtimeJobClassName);
+        LOG.info("Invoking destroy() on {}", dataprocEnvClassName);
```
Comment (CuriousVini) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocJobMain.java:143]: why is this change needed?

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeMonitorTestUtil.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright © 2014 Cask Data, Inc.
```
Comment (CuriousVini) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeMonitorTestUtil.java]: 2021

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java [hunk 9]
```diff
@@ -73,48 +73,49 @@
...
-  private boolean isValid(ProgramRunId programRunId) throws IOException, UnauthorizedException {
+  private RunRecordDetail runRecord(ProgramRunId programRunId) throws IOException, UnauthorizedException,
+    NotFoundException {
     RunRecordDetail runRecord = TransactionRunners.run(txRunner, context -> {
       return AppMetadataStore.create(context).getRun(programRunId);
     }, IOException.class);
 
     if (runRecord != null) {
-      return !runRecord.getStatus().isEndState();
+      return runRecord;
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java]: What is the impact if we remove this check? Originally this method was being used to validate the record status.

**Review by CuriousVini — COMMENTED** [review]
Thanks, Took a first pass, A few things to consider:
1. we should have jira for each PR change
2. Could you also add description on the PR on what is the change about?
3. If the runtime handler response is changing, could we have a unit test with that change being reflected?

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocJobMain.java (hunk 7 above)
Comment (snehalakshmisha) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocJobMain.java:143]: This line was printed as `2021-12-10 03:06:51,302 INFO runtimejob.DataprocJobMain: Invoking destroy() on io.cdap.cdap.internal.app.runtime.distributed.runtimejob.DefaultRuntimeJob`
destroy() method does not exist in https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJob.java but is actually in dataprocenvclass - https://github.com/cdapio/cdap/blob/develop/cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocRuntimeEnvironment.java#L102

File: cdap-proto/src/main/java/io/cdap/cdap/proto/ProgramRunStatus.java (hunk 6 above)
Comment (snehalakshmisha) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/ProgramRunStatus.java]: As discussed offline, these changes will be made in one of the following PRs

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java (hunk 4 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java:102]: Actually, I am keeping the current behavior intact. The existing code also throws a ServiceUnavailbleException - https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java#L90 - so there is no change in the method contract with new changes

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java [hunk 10]
```diff
@@ -115,6 +123,16 @@ public void sendMessages(ProgramRunId programRunId,
       }
 
       throwIfError(programRunId, urlConn);
+      // fetch grace period for stopping
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java]: Should we have this as part of a future PR?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeHandler.java [hunk 11]
```diff
@@ -143,15 +150,30 @@ public BodyConsumer writeMessages(HttpRequest request, HttpResponder responder,
...
       throw new UnauthorizedException("Access denied for topic " + topic);
     }
 
     TopicId topicId = NamespaceId.SYSTEM.topic(topic);
     if (topic.startsWith(logsTopicPrefix)) {
-      return new MessageBodyConsumer(topicId, logProcessor::process);
+      return new MessageBodyConsumer(topicId, logProcessor::process, runRecordStatus);
+    }
+
+    // after 10 mins add stopping state and timeout here
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeHandler.java]: the below lines are only for testing and will be removed before merging

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java [hunk 12]
```diff
@@ -73,49 +74,76 @@
     long pollTimeMillis = cConf.getLong(Constants.RuntimeMonitor.POLL_TIME_MS);
     this.programRunsCache = CacheBuilder.newBuilder()
       .expireAfterWrite(pollTimeMillis, TimeUnit.MILLISECONDS)
-      .build(new CacheLoader<ProgramRunId, Boolean>() {
+      .build(new CacheLoader<ProgramRunId, ImmutablePair<Boolean, RunRecordStatus>>() {
+        /**
+         * {@inheritDoc}
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java]: we can remove this

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java [hunk 13]
```diff
@@ -73,49 +74,76 @@
...
-      return AppMetadataStore.create(context).getRun(programRunId);
-    }, IOException.class);
-
-    if (runRecord != null) {
-      return !runRecord.getStatus().isEndState();
-    }
-    // If it is not found in the local store, which should be very rare, try to fetch the run record remotely.
+  private ImmutablePair<Boolean, RunRecordStatus> getRunRecordStatusForProgramsInNonEndState(ProgramRunId programRunId)
+    throws IOException,
+    UnauthorizedException {
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java]: move it to previous line

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java (hunk 10 above)
Comment (CuriousVini) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java]: should it be part of the response where stopping status is sent from runtime handler? That way we dont need to wait for 2 calls for getting all the details

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeHandler.java (hunk 11 above)
Comment (CuriousVini) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeHandler.java]: Usually we only keep the code changes in the PR that needs to be reviewed and merge. I think its safe to remove it from this PR

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeHandler.java [hunk 14]
```diff
@@ -244,15 +266,20 @@ public void handleError(Throwable cause) {
     private final List<byte[]> payloads;
     private ByteBuffer payload;
     private long items;
+    private final RunRecordStatus runRecordStatus;
+    private static final Gson GSON = new GsonBuilder()
+      .setPrettyPrinting()
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeHandler.java]: we can remove this

File: cdap-proto/src/main/java/io/cdap/cdap/proto/ProgramRunStatus.java (hunk 6 above)
Comment (CuriousVini) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/ProgramRunStatus.java]: can you file a jira and add a TODO here for tracking?

**Review by CuriousVini — COMMENTED** [review]
a few comments

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java (hunk 10 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClient.java]: As discussed offline, keeping the below lines. Also, the stopping status will be part of the response from runtime handler

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java [hunk 15]
```diff
@@ -144,6 +144,12 @@ long getProgramFinishTime() {
     return programFinishTime;
   }
 
+  public void getStoppingCommandWithGracefulPeriod() throws IOException, BadRequestException {
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java]: Remove this

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+    this.programRunStatus = programRunStatus;
+    this.message = message;
+  }
+
+  public ProgramRunStatus getProgramRunStatus() {
+    return programRunStatus;
+  }
+
+  @Nullable
+  public String getMessage() {
```
Comment (snehalakshmisha) [inline cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java]: Use json and call it payload

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/store/RunRecordDetail.java [hunk 17]
```diff
@@ -67,12 +67,13 @@
   private transient volatile Map<String, String> userArgs;
 
   private RunRecordDetail(ProgramRunId programRunId, long startTs, @Nullable Long runTs, @Nullable Long stopTs,
-                          @Nullable Long suspendTs, @Nullable Long resumeTs,
+                          @Nullable Long suspendTs, @Nullable Long resumeTs,  @Nullable Long stoppingTs,
```
Comment (snehalakshmisha) [inline cdap-common/src/main/java/io/cdap/cdap/internal/app/store/RunRecordDetail.java:70]: Add graceful shutdown period as well

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java [hunk 18]
```diff
@@ -73,49 +74,72 @@
...
       });
   }
 
   @Override
-  public void validate(ProgramRunId programRunId, HttpRequest request) throws BadRequestException {
+  public RunRecordStatus checkProgramRunStatus(ProgramRunId programRunId, HttpRequest request) throws
+    BadRequestException {
     accessEnforcer.enforce(programRunId, authenticationContext.getPrincipal(), StandardPermission.GET);
-    boolean exists;
+    ImmutablePair<Boolean, RunRecordStatus> pair;
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java]: Look at just the second object

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidatorTest.java [hunk 19]
```diff
@@ -226,11 +231,15 @@ public void testFetcher() throws BadRequestException {
...
+      HttpVersion.HTTP_1_1, HttpMethod.GET, "/"));
+    Assert.assertEquals(programRunStatus, runRecordStatus.getProgramRunStatus());
 
     // The second call will hit the runtime store, so it shouldn't matter what the run record fetch returns
     runRecordFetcher.setRunRecord(null);
-    validator.validate(programRunId, new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/"));
+    runRecordStatus = validator.checkProgramRunStatus(programRunId, new DefaultHttpRequest(HttpVersion.HTTP_1_1,
+                                                                                           HttpMethod.GET, "/"));
+    Assert.assertEquals(programRunStatus, runRecordStatus.getProgramRunStatus());
   }
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidatorTest.java:246]: Add a new unit test for stopping with payload

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java (hunk 16 above)
Comment (snehalakshmisha) [reply cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java]: Bytes payload and stopping timestamp in future

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeHandler.java [hunk 20]
```diff
@@ -17,6 +17,8 @@
 package io.cdap.cdap.internal.app.runtime.monitor;
```
Comment (snehalakshmisha) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeHandler.java:17]: Calculate stopping future time here

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidatorTest.java (hunk 19 above)
Comment (snehalakshmisha) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidatorTest.java:246]: Will be part of https://cdap.atlassian.net/browse/CDAP-18743

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java [hunk 21]
```diff
@@ -73,51 +75,66 @@
...
-          return isValid(programRunId);
+        public Optional<RunRecordStatus> load(ProgramRunId programRunId) throws IOException, UnauthorizedException {
+          return Optional.ofNullable(getRunRecordStatusForProgramsInNonEndState(programRunId));
         }
       });
   }
 
   @Override
-  public void validate(ProgramRunId programRunId, HttpRequest request) throws BadRequestException {
+  public RunRecordStatus checkProgramRunStatus(ProgramRunId programRunId, HttpRequest request) throws
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java]: better call it getProgramRunStatus

Comment (CuriousVini) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java]: nit: move throws to next line along with exception for readability

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+import io.cdap.cdap.api.data.format.UnexpectedFormatException;
+import io.cdap.cdap.proto.ProgramRunStatus;
+
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+/**
+ * Used to send information about program status from RuntimeHandler to RuntimeClient.
+ */
+public class RunRecordStatus {
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java]: this class carries more information than just program status. May be worth renaming it

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+  private final ProgramRunStatus programRunStatus;
+  @Nullable
+  private final String payload;
+
+  private RunRecordStatus(ProgramRunStatus programRunStatus, String payload) {
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java]: should tag payload to be nullable in method param

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+      return true;
+    }
+
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    RunRecordStatus that = (RunRecordStatus) o;
+    return programRunStatus.equals(that.programRunStatus) &&
+      payload.equals(that.payload);
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java]: does this throw NPE if payload is null?

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+  public static Builder builder() throws UnexpectedFormatException {
+    return new Builder();
+  }
+
+  /**
+   * Builder for creating a {@link RunRecordStatus}.
+   */
+  public static class Builder {
+    private ProgramRunStatus programRunStatus;
+    private String payload;
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java]: annotate this as nullable

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java [hunk 26]
```diff
@@ -73,51 +75,66 @@
...
+      return runRecord.getStatus().isEndState() ? null : getValidRunRecordStatus(runRecord);
     } catch (NotFoundException e) {
-      return false;
+      return null;
     }
   }
 
+  private RunRecordStatus getValidRunRecordStatus(RunRecordDetail runRecord) {
+    ProgramRunStatus programRunStatus = runRecord.getStatus();
+    RunRecordStatus.Builder runRecordStatusBuilder = RunRecordStatus.builder().setProgramRunStatus(programRunStatus);
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java]: If both parameters are set at the same time, do you think builder pattern is needed in this case?

**Review by CuriousVini — COMMENTED** [review]
a few comments

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java (hunk 24 above)
Comment (snehalakshmisha) [reply cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java]: Oops, it does! Changed it

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java (hunk 25 above)
Comment (snehalakshmisha) [reply cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RunRecordStatus.java]: Got rid of builder as discussed

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeMonitorTestUtil.java [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package io.cdap.cdap.internal.app.runtime.monitor;
+
+import io.cdap.cdap.proto.ProgramRunStatus;
+
+/**
+ * Utility class for runtime monitor tests
+ */
+public final class RuntimeMonitorTestUtil {
```
Comment (CuriousVini) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeMonitorTestUtil.java]: If this only has one - one line - method then we can just create this object where its needed and remove this class?

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/ProgramRunStatusInfo.java [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+
+import io.cdap.cdap.proto.ProgramRunStatus;
+
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+/**
+ * Used to send information about program status from RuntimeHandler to RuntimeClient.
+ */
+public class ProgramRunStatusInfo {
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/ProgramRunStatusInfo.java]: May be call it `ProgramRunInfo`? Since it contains status and timeout information about a specific program run.

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/ProgramRunStatusInfo.java [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(getProgramRunStatus(), getPayload());
+  }
+
+  @Override
+  public String toString() {
+    return com.google.common.base.Objects.toStringHelper(this)
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/ProgramRunStatusInfo.java]: Why do we need to use com.google class? Can we simply use java's string builder OR follow existing pattern? 
For example: ETLBatchConfig#toString()

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/store/RunRecordDetail.java [hunk 30]
```diff
@@ -172,6 +175,8 @@ public String toString() {
       .add("stopTs", getStopTs())
       .add("suspendTs", getSuspendTs())
       .add("resumeTs", getResumeTs())
+      .add("stoppingTs", getStoppingTs())
+      .add("gracefulStoppingTimeout", getGracefulStoppingTimeout())
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/internal/app/store/RunRecordDetail.java]: May be just call it stopping timeout?

**Review by CuriousVini — COMMENTED** [review]
a few minor comments

File: cdap-proto/src/main/java/io/cdap/cdap/proto/RunRecord.java [hunk 31]
```diff
@@ -270,6 +301,16 @@ public T setResumeTime(Long resumeTs) {
       return (T) this;
     }
 
+    public T setStoppingTime(Long stoppingTs) {
+      this.stoppingTs = stoppingTs;
+      return (T) this;
+    }
+
+    public T setStoppingTimeout(Long stoppingTimeout) {
```
Comment (CuriousVini) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/RunRecord.java]: shouldnt this be stoppingTimeoutTs?

File: cdap-proto/src/main/java/io/cdap/cdap/proto/ProgramRunStatus.java (hunk 6 above)
Comment (CuriousVini) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/ProgramRunStatus.java]: missed?

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/ProgramRunInfo.java [hunk 32]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(getProgramRunStatus(), getPayload());
+  }
+
+  @Override
+  public String toString() {
+    final StringBuilder sb = new StringBuilder();
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/ProgramRunInfo.java]: no need to make local variables final

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java [hunk 33]
```diff
@@ -73,51 +75,64 @@
...
+        /**
+         * For a programRunId this cache stores a {@link ProgramRunInfo} object only for valid programRunIds,
+         * i.e. this value will be Optional.empty() if the programRunId is invalid.
+         */
         @Override
-        public Boolean load(ProgramRunId programRunId) throws IOException, UnauthorizedException {
-          return isValid(programRunId);
+        public Optional<ProgramRunInfo> load(ProgramRunId programRunId)
+          throws IOException, UnauthorizedException {
+          return Optional.ofNullable(getRunRecordStatusForProgramsInNonEndState(programRunId));
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java]: nit: may be just return optional from the method itself so you dont have to wrap it here with optional

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeRequestValidator.java [hunk 34]
```diff
@@ -27,12 +27,15 @@
...
+   * Otherwise, returns the program status and any information associated with it.
    *
    * @param programRunId the program run id where the request is coming from
    * @param request the http request from the program runtime
    * @throws BadRequestException if the request is not valid
    * @throws AccessException if the request doesn't have a valid Authorization header
+   * @return RunRecordDetail if the programRunId was found
    */
-  void validate(ProgramRunId programRunId, HttpRequest request) throws BadRequestException, AccessException;
+  ProgramRunInfo getProgramRunStatus(ProgramRunId programRunId, HttpRequest request) throws BadRequestException,
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeRequestValidator.java]: move throws below for readability.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeRequestValidator.java [hunk 35]
```diff
@@ -27,12 +27,15 @@
...
   /**
-   * Validates the request.
+   * Validates that the program is not in end state, if so throws BadRequestException.
+   * Otherwise, returns the program status and any information associated with it.
    *
    * @param programRunId the program run id where the request is coming from
    * @param request the http request from the program runtime
    * @throws BadRequestException if the request is not valid
    * @throws AccessException if the request doesn't have a valid Authorization header
+   * @return RunRecordDetail if the programRunId was found
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeRequestValidator.java]: Obsolete javadoc?

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidatorTest.java (hunk 19 above)
Comment (CuriousVini) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidatorTest.java:246]: add a TODO in code with jira number

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/ProgramRunInfo.java (hunk 32 above)
Comment (CuriousVini) [reply cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/ProgramRunInfo.java]: Also intellij coding style should recommend using following:
```
return "ProgramRunInfo" +
      "{programRunStatus=" + programRunStatus +
      ", payload='" + payload +
      '}';

**Review by CuriousVini — COMMENTED** [review]
a few comments

File: cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/ProgramRunInfo.java (hunk 32 above)
Comment (snehalakshmisha) [reply cdap-common/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/ProgramRunInfo.java]: Refactored. IntelliJ did not recommend anything, maybe I am missing adding the lint settings.
Also, I noticed a mix of implementations for toString() with StringBuilder as well as guava's Objects in multiple files in this repo.

**Review by CuriousVini — COMMENTED** [review]
LGTM

**Review by CuriousVini — APPROVED** [review]
